docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,333 @@
+To make this concrete:
+```julia
+y = f(x)  # primal program
+x̄ = pullback_at(f, x, y, ȳ)
+```
+
+Let's write one:
+```julia
+y = sin(x)
+pullback_at(::typeof(sin), x, y, ȳ) = ȳ * cos(x)
+```
+
+Great. So far so good.
+As a short exercise the reader might like to implement the one for the [logistic sigmoid](https://en.wikipedia.org/wiki/Logistic_function#Derivative).
+It also works without issue.
+#TODO should I actually just include logistic sigmoid a second example though out, and write its code next the the code for `sin`?
+
+
+Now let's consider why we implement `rrules` like this in the first place.
+One key reason, [^3] is to allow us to insert our domain knowledge to do better than the AD would do just by breaking everything down into `+` and `*` etc.
+What insights do we have about `sin` and `cos`?
+Here is one:
+```julia
+julia> @btime sin(x) setup=(x=rand());
+  3.838 ns (0 allocations: 0 bytes)
+
+julia> @btime cos(x) setup=(x=rand());
+  4.795 ns (0 allocations: 0 bytes)
+
+julia> 3.838 + 4.795
+8.633
+
+julia> @btime sincos(x) setup=(x=rand());
+  6.028 ns (0 allocations: 0 bytes)
+```
+It is \~30%[^4] faster to compute `sin` and `cos` at the same time via `sincos` than it is to compute them one after the other.
+How can we incorporate this insight into our system?
+We know that we can compute the `cos(x)` at the same time as the `sin(x)` is computed in primal, because it only depends on `x` --- we don't need to know `ȳ`.
+but there is no-where to put it that is accessible both to the primal pass, and the gradient pass code.
+
+What if we introduced some variable called `intermediates` that is also record onto the tape during the primal pass?
+We would need to be able to modify the primal pass to do this, so we can actually put the data into the `intermediates`.
+So we will introduce a function: `augmented_primal`, that will return the primal output, plus the `intermediates` that we want to reuse in the gradient pass.
+Then we will make our AD system replace calls to the primal with calls to the `augmented_primal` of the primal function; and take care of all the bookkeeping.
+So that would look like:
+```julia
+y = f(x)  # primal program
+y, intermediates = augmented_primal(f, x)
+x̄ = pullback_at(f, x, y, ȳ, intermediates)
+```
+
+Let's try writing this now:
+```julia
+y = sin(x)
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, (; cx=cx)  # use a NamedTuple for the intermediates
+end
+
+pullback_at(::typeof(sin), x, y, ȳ, intermediates) = ȳ * intermediates.cx
+```
+
+Cool!
+That lets us do what we wanted.
+We net deceased the time it takes to run the primal and gradient passes.
+We have now demonstrated the title question of why we needed to be able to modify the primal pass.
+We will go into that more later, and have some more examples of use.
+But first lets continue to see how we go from that `augmented_primal` _ `pullback_at` to [`rrule`](@ref).
```

Comment:
```suggestion
But first let's continue to see how we go from that `augmented_primal` to `pullback_at` to [`rrule`](@ref).
```

Reply:
I'm not sure that what the underscore here was for

Reply:
weird typo